Submenu to select push-application

calixtus opened the pull request:
Since the migration to JavaFX, i saw no way to change the push-application. There were some pull-request in the past, trying to fix this, but they all were forgotten undone. So this is my shot, trying to implement a very simple way to change the push-application.

fixes #674 closes #3942 and closes #4368 closes #5024

~~(Apparently, I forgot the fetch before I opened the new branch. Please ignore the old superfluous commits, my changes start with [95f8668](https://github.com/JabRef/jabref/pull/4991/commits/95f86688d18e9081a890b9a0f920e65e72168b3f)~~

~~Attention: It's still WIP,  it's not done yet.
The caption of the menu is temporary, there is no l10n, Toolbarbutton is not updated and another small Problem is left, i could not figure out yet and I am hoping for your support: I was not able to change the caption of the PushAppplicationMenuItem, since it's bound somehow to another variable. Do you have any Idea, how i can change this label?~~

<img width="370" alt="tools-push" src="https://user-images.githubusercontent.com/50491877/58163345-27657a80-7c84-11e9-8946-d075a0316311.png">
----

- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

calixtus commented on src/main/java/org/jabref/gui/push/PushToApplicationMenuAction.java:
```diff
@@ -0,0 +1,61 @@
+    public void execute() {
+        Globals.prefs.put(JabRefPreferences.PUSH_TO_APPLICATION, application.getApplicationName());
+        pushToApplicationAction.updateApplication(application);
+        //pushToApplicationMenuItem.setText(pushToApplicationAction.getActionInformation().getText());
```
This does not work. Any ideas?

Siedlerchr replied:
Hm, do you get any error message?

calixtus replied:
This code works fine so far, I just put this last line in comment to show what I'm going to do. When I compile "live" with this last line, JabRef displays a runtime error in "execution".

Siedlerchr replied:
Please post the stacktrace and the error message. Something like a bound value cannot be set?
That would imply that there is somehow a Binding on that property

calixtus replied:
java.lang.RuntimeException: MenuItem.text : A bound value cannot be set.
	at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:140)
	at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:49)
	at javafx.scene.control.MenuItem.setText(MenuItem.java:257)
	at org.jabref.gui.push.PushToApplicationMenuAction.execute(PushToApplicationMenuAction.java:59)
	at org.jabref.gui.actions.JabRefAction.lambda$new$30(JabRefAction.java:29)
	at org.controlsfx.control.action.Action.handle(Action.java:419)
	at org.controlsfx.control.action.Action.handle(Action.java:64)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1405)
	at com.sun.javafx.scene.control.skin.ContextMenuContent$MenuItemContainer.lambda$createChildren$343(ContextMenuContent.java:1358)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$358(GlassViewEventHandler.java:432)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:410)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$152(WinApplication.java:177)
	at java.lang.Thread.run(Thread.java:748)

Siedlerchr replied:
As I expected. There is a binding on the MenuItem.textProperty() (Binding is javafx (bi)directional databinding on properties). I think this could be related to the PR from @tobiasdiez  in which he implemeted the disabling of menu items.

calixtus replied:
Mh, I have no experience with bindings, but this sounds like, i should better not change something, which is still WiP by @tobiasdiez. As this is just eye-candy, in can make this label static and freeze also the symbol in the Toolbar. (I was surprised, it actually changes to the proper symbol, if you restart JabRef)

Siedlerchr replied:
Actually it's very practical: https://docs.oracle.com/javase/8/javafx/properties-binding-tutorial/binding.htm
And also see our wiki for some general javafx explainations/tips
https://github.com/JabRef/jabref/wiki/Code---JavaFX

Siedlerchr commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -760,6 +765,23 @@ private MenuBar createMenu() {
+        RadioMenuItem pushApplication;
+        PushToApplicationMenuAction pushToApplicationMenuAction;
+
+        for (PushToApplication application : JabRefFrame.this.getPushApplications().getApplications()) {
```
this.getPushApplications?

calixtus replied:
Thanks, I copied this for-loop out of externalTab and did not simplify it yet.

calixtus commented:
Any suggestions for the label of the submenu? Push-application sounds so technical.

calixtus commented:
So, theoretically this would be ready to review now...
If I hadn't found a bug in another place.

<img width="148" alt="button-too-small" src="https://user-images.githubusercontent.com/50491877/58192366-dcb62380-7cc0-11e9-95dd-ba1e50d0aa63.png">

The button is shown too small after updating. And I know why.

I use the method configureIconButton in the ActionFactory to unbind and refresh the ToolBarButton. 
Obviously someone cheated, when he "fixed" an earlier bug with the iconsize in the entry-editor... ;-)

<img width="313" alt="icon-cheater" src="https://user-images.githubusercontent.com/50491877/58192547-433b4180-7cc1-11e9-9a01-d67a7b381952.png">

calixtus commented:
The other Pull Request should fix the button-size-issue. Thats all, im ready to review.

Siedlerchr commented:
Maybe wording for Push applications
Push to External editor? 

Regarding your superfluous commits,if you do a git rebase on the upstream/master (after fetch) then only the new changes should be visible

calixtus commented:
I think, that didn't work... I stop now rebasing, it's getting out of hand.
The code is the same, i checked twice.

matthiasgeiger commented:
One question: Which IDE or Git tool are you using? 
Special stuff like rebasing is very well supported in IntelliJ (I'm also not capable of rebasing using just the command line ;-))

... passend dazu: https://classicprogrammerpaintings.com/post/142586036029/junior-programmer-learns-git-rebase

calixtus commented:
It's IntelliJ IDEA CE, i wanted to make sure, if i pass commits, they are formatted right, and i read in the documentation, that you recommend this. But i just start to get used to it. In my early days i used eclipse. I pushed with the commandline tool which is probably the cause for this mess.

calixtus commented:
Cleaned up the commit history, the rebase-monster is finally gone.
As @Siedlerchr pointed out a few days ago: "Nothing git magic can't fix": I sync'd my local master branch with upstream/master, hard-reset my working-branch history to my last real commit, rebased on master and force-pushed to github.

Siedlerchr commented:
So just one question to help me understand the changes of this a bit better, whenever I change the push application in the tools submenu the icon button toolbar changes to the selected application?

codewise lgtm so far

calixtus commented:
The RadioMenuItems change
* the setting in preferences and
* the Action stored in the Manager
and they reconfigure
* the MenuItem and
* the ToolBarButton.

tobiasdiez commented on src/main/java/org/jabref/gui/JabRefFrame.java:
```diff
@@ -759,7 +767,30 @@ private MenuBar createMenu() {
+        final String activePushToApplicationName = Globals.prefs.get(JabRefPreferences.PUSH_TO_APPLICATION);
+
+        final ToggleGroup pushToToggleGroup = new ToggleGroup();
+        RadioMenuItem pushToApplication;
```
Please move the declaration of `pushToApplication` and `pushToApplicationMenuAction` down to the body of the for loop (because otherwise one would expect that you need these variables also after the for-loop).

calixtus replied:
done

tobiasdiez commented on src/main/java/org/jabref/gui/push/PushToApplicationAction.java:
```diff
@@ -25,25 +25,37 @@
 
     public Action getActionInformation() {
-        return new Action() {
+        return action;
```
What was the reason to change `getActionInformation` (i.e. instead of returning a new instance of the action, keep one and update it)?

calixtus replied:
It keeps the toolBarButton and the menuItem in sync with the same action.

calixtus replied:
You were absolutely right. It was an artifact of an earlier version, when I needed it as a persistent action. I reverted it and it still works fine.

calixtus commented:
> Thanks for your contribution. Code-wise the changes look good to me and I've only a few a few suggestions.
> Although its a huge improvement in terms of usability, I'm not sure if the solution to set the push to application through a separate menu entry is intuitive to most users. I would prefer if the application is set in the preferences instead of via a menu item. @JabRef/developers opinions?

Shouldn't be a big deal to put it into the preferences. I'll look into it tonight. I going to create a second pull request for the alternative.

calixtus commented:
> Thanks for your contribution. Code-wise the changes look good to me and I've only a few a few suggestions.
> Although its a huge improvement in terms of usability, I'm not sure if the solution to set the push to application through a separate menu entry is intuitive to most users. I would prefer if the application is set in the preferences instead of via a menu item. @JabRef/developers opinions?

I put an alternative proposal in #5024
